### PR TITLE
Combine evaluate_stateful and evaluate_inside_range

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Which issue does this PR close?

Related to #5781 

# Rationale for this change

# What changes are included in this PR?
This PR implements the proposal in the [#PR12](https://github.com/alamb/arrow-datafusion/pull/12)
- `evaluate_stateful` and `evaluate_inside_range` are combined under single `evaluate` methods
- `uses_window_frame` and `supports_bounded_execution` methods are moved from `BoundedWindowFunctionExpr` to `PartitionEvaluator`.

With the changes in this PR new Evaluators can be implemented according to table below.
|uses_window_frame|supports_bounded_execution|function_to_implement|
|---|---|----|
|false|false|`evaluate_all` (if we were to implement `PERCENT_RANK` it would end up in this quadrant, we cannot produce any result without seeing whole data)|
|false|true|`evaluate` (optionally can also implement `evaluate_all` for more optimized implementation. However, there will be default implementation that is suboptimal) . If we were to implement `ROW_NUMBER` it will end up in this quadrant. Example `OddRowNumber` showcases this use case|
|true|false|`evaluate` (I think as long as `uses_window_frame` is `true`. There is no way for `supports_bounded_execution` to be false). I couldn't come up with any example for this quadrant |
|true|true|`evaluate`. If we were to implement `FIRST_VALUE`, it would end up in this quadrant|.

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/window/partition_evaluator.rs [hunk 1]
```diff
@@ -132,23 +140,40 @@ pub trait PartitionEvaluator: Debug + Send {
...
+        // If user wants to implement more efficient version, this method should be overwritten
+        // Default implementation may behave suboptimally (For instance `NumRowEvaluator` overwrites it)
+        if !self.uses_window_frame() && self.supports_bounded_execution() {
+            let res = (0..num_rows)
+                .map(|_idx| self.evaluate(values, &Range { start: 0, end: 1 }))
+                .collect::<Result<Vec<_>>>()?;
+            ScalarValue::iter_to_array(res.into_iter())
+        } else {
+            Err(DataFusionError::NotImplemented(
+                "evaluate_all is not implemented by default".into(),
```
Comment (alamb) [inline datafusion/physical-expr/src/window/partition_evaluator.rs:158]:
```suggestion
                format!("evaluate_all is not implemented for {} when using window frames", self.name()),
```

I think the error could be more helpful, but I can also add that as a follow on PR

**Review by alamb — APPROVED** [review]
This makes sense to me @mustafasrepo  and is a very nice cleanup. Thank you!

File: datafusion/physical-expr/src/window/built_in_window_function_expr.rs [hunk 2]
```diff
@@ -86,27 +86,10 @@ pub trait BuiltInWindowFunctionExpr: Send + Sync + std::fmt::Debug {
...
-    /// implement [`PartitionEvaluator::evaluate_inside_range`]
-    fn uses_window_frame(&self) -> bool {
-        false
-    }
-
     /// Can this function be evaluated with (only) rank
     ///
     /// If `include_rank` is true, then [`Self::create_evaluator`] must
-    /// implement [`PartitionEvaluator::evaluate_with_rank`]
+    /// implement [`PartitionEvaluator::evaluate_with_rank_all`]
```
[the remark this answers is not in the document]
Comment (mustafasrepo) [reply datafusion/physical-expr/src/window/built_in_window_function_expr.rs]: For consistency, I have moved `include_rank` flag to `PartitionEvaluator` also.

File: datafusion/physical-expr/src/window/rank.rs [hunk 3]
```diff
@@ -115,12 +115,6 @@ pub(crate) struct RankEvaluator {
 }
 
 impl PartitionEvaluator for RankEvaluator {
-    fn get_range(&self, idx: usize, _n_rows: usize) -> Result<Range<usize>> {
```
Comment (alamb) [inline datafusion/physical-expr/src/window/rank.rs:126]: ❤️

**Review by alamb — APPROVED** [review]
Looks great to me -- thanks again @mustafasrepo

**Pull request merged**